### PR TITLE
Add mechanism for application/container migration in cluster

### PR DESCRIPTION
**Pull request by naiming-zededa**
- convert native container pod into replicatset
- use EdgeNodeInfo for getting the node name and uuid for domainmgr
- domainmgr prevent delete app due to app not showing up, it's a replicaset, domainmgr should not worry about deleting
- try to fillin the volumemgr Volume Reference name for container, which needs to be kept all the way through volumemgr -> zedmanager -> domainmgr
- some temp code on cluster pubsub
- generate uniform random Mac address across cluster nodes for App

### CODE REVIEW

**Comment by andrewd-zededa** [comment]
@naiming-zededa where is the vmi replicaset created?

**Comment by naiming-zededa** [comment]
@andrewd-zededa the vmi replicaset is created in CreateReplicaSetConfig() function, eventually with
```
	meta := vmiMetaData{
		replica:  replicaSet,
		mtype:    IsMetaReplicaSet,
		name:     kubeName,
		domainID: int(rand.Uint32()),
	}
	ctx.vmiList[domainName] = &meta
```

**Comment by andrewd-zededa** [comment]
@naiming-zededa Where will the designated node id be set?  I'm thinking about the descheduler interaction.

**Comment by andrewd-zededa** [comment]
Ok, that should match the descheduler direction.

File: pkg/pillar/cmd/zedagent/parseconfig.go [hunk 1]
```diff
@@ -2376,8 +2376,7 @@ func isOverlayNetworkInstance(netInstEntry *zconfig.NetworkInstanceConfig) bool
 	return netInstEntry.InstType == zconfig.ZNetworkInstType_ZnetInstMesh
 }
 
-func parseAppNetAdapterConfigEntry(
-	cfgApp *zconfig.AppInstanceConfig,
+func parseAppNetAdapterConfigEntry(cfgApp *zconfig.AppInstanceConfig,
```
Comment (andrewd-zededa) [inline pkg/pillar/cmd/zedagent/parseconfig.go:2379]: Looks like nothing changed here, not a big deal

File: pkg/pillar/cmd/zedkube/applogs.go [hunk 2]
```diff
@@ -50,11 +56,14 @@ func collectAppLogs(ctx *zedkubeContext) {
...
+		if aiconfig.FixedResources.VirtualizationMode != types.NOHYPER {
+			continue
+		}
 		if aiconfig.DesignatedNodeID != uuid.Nil && aiconfig.DesignatedNodeID.String() != ctx.nodeuuid { // For now, only check DNiD, need to add migration part
 			continue
 		}
-		contName := base.GetAppKubeName(aiconfig.DisplayName, aiconfig.UUIDandVersion.UUID)
-
+		kubeName := base.GetAppKubeName(aiconfig.DisplayName, aiconfig.UUIDandVersion.UUID)
+		contName := kubeName
```
Comment (andrewd-zededa) [inline pkg/pillar/cmd/zedkube/applogs.go:66]: Do we need both here?

File: pkg/pillar/dpcreconciler/linux.go [hunk 3]
```diff
@@ -832,8 +832,18 @@ func (r *LinuxDpcReconciler) updateCurrentRoutes(dpc types.DevicePortConfig) (ch
...
+				_, tmpRoute, err := net.ParseCIDR(kubePodCIDR)
+				if err != nil {
+					r.Log.Errorf("updateCurrentRoutes: failed to parse CIDR %s: %v",
+						kubePodCIDR, err)
+					continue
+				}
+				routeTmp.Dst = tmpRoute
 				currentRoutes.PutItem(linux.Route{
-					Route:         rt.Data.(netlink.Route),
+					//Route:         rt.Data.(netlink.Route),
```
Comment (andrewd-zededa) [inline pkg/pillar/dpcreconciler/linux.go:845]: The commented out old version could be found in git log instead, do we need it here?

File: pkg/pillar/types/contenttreetypes.go [hunk 4]
```diff
@@ -31,6 +31,8 @@ type ContentTreeConfig struct {
 	CustomMeta        string
 	// XXX DesignatedNodeID
 	DesignatedNodeID uuid.UUID
```
Comment (andrewd-zededa) [inline pkg/pillar/types/contenttreetypes.go:33]: A comment about its usage only in kubevirt could be helpful to have for others.

**Review by andrewd-zededa — COMMENTED** [review]
I think in general it looks ok, are the DNid fields available in api to exercise this code yet?

**Comment by naiming-zededa** [comment]
> I think in general it looks ok, are the DNid fields available in api to exercise this code yet?

In this code, the DomainConfig structure prepared by 'zedmanager' will set 'IsDNidNode' flag. Currently is used to prevent non-DNiD node to delete the app. When creating, this code does not even check this flag, it just uses it's own ctx.nodeName for this.

File: pkg/pillar/cmd/zedagent/parseconfig.go (hunk 1 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedagent/parseconfig.go:2379]: right, i used to add something, but removed that. need to realign the lines

File: pkg/pillar/dpcreconciler/linux.go (hunk 3 above)
Comment (naiming-zededa) [reply pkg/pillar/dpcreconciler/linux.go:845]: i'll remove

File: pkg/pillar/types/contenttreetypes.go (hunk 4 above)
Comment (naiming-zededa) [reply pkg/pillar/types/contenttreetypes.go:33]: ok.

File: pkg/pillar/cmd/zedkube/applogs.go (hunk 2 above)
Comment (naiming-zededa) [reply pkg/pillar/cmd/zedkube/applogs.go:66]: i think this is the case, the contName can change, the kubeName can be the prefix of a real name:
```
		for _, pod := range pods.Items {
			if strings.HasPrefix(pod.ObjectMeta.Name, kubeName) {
				contName = pod.ObjectMeta.Name
				break
			}
		}
```

**Review by andrewd-zededa — APPROVED** [review]
Ok, LGTM.  I might need to rebase PR #344 a bit, I added EdgeNodeInfo Create/Modify handler so I can do uncordon in a goroutine when its ready.

**Comment by naiming-zededa** [comment]
@andrewd-zededa go ahead rebase your PR 344, and i have added some changes using the nodeName already in drain function.

**Pull request merged**